<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# **Hands-on Lab: Interactive Visual Analytics with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:

*   **TASK 1:** Mark all launch sites on a map
*   **TASK 2:** Mark the success/failed launches for each site on the map
*   **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [50]:
# Install deps in both environments (JupyterLite or local/Colab)
try:
    import piplite  # only exists in JupyterLite (Pyodide)
    # JupyterLite precisa de await:
    await piplite.install(['folium'])
    # pandas geralmente já vem; instale só se faltar:
    try:
        import pandas as pd  # noqa
    except Exception:
        await piplite.install(['pandas'])
except Exception:
    # Ambiente local/Colab: use pip normal
    import sys, subprocess
    subprocess.check_call([sys.executable, "-m", "pip", "install", "folium"])

# Imports
import folium
from folium.plugins import MarkerCluster, MousePosition, MiniMap
from folium.features import DivIcon
import pandas as pd

In [51]:
import folium
import pandas as pd

In [52]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/labs/v4/DV0101EN-Exercise-Generating-Maps-in-Python.ipynb)


In [55]:
## Task 1: Mark all launch sites on a map


First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site.


In [64]:
# Download and read the `spacex_launch_geo.csv`
# 1) Load the CSV (works in both JupyterLite and local environments)
import io, pandas as pd
URL = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv"

try:
    from js import fetch                          # JupyterLite
    resp = await fetch(URL)
    b = await resp.arrayBuffer()
    df = pd.read_csv(io.BytesIO(bytearray(b)))
except Exception:
    # Ambiente local/Colab
    try:
        df = pd.read_csv(URL)
    except Exception:
        import urllib.request
        urllib.request.urlretrieve(URL, "spacex_launch_geo.csv")
        df = pd.read_csv("spacex_launch_geo.csv")

# 2) Check column names
print(df.columns.tolist())

# 3) Normalize column names and create the expected subset
cols = [c.lower().strip() for c in df.columns]
df.columns = cols

# Accept either 'lat/long' or 'latitude/longitude'
if "lat" in df.columns and "long" in df.columns:
    lat_col, lon_col = "lat", "long"
elif "latitude" in df.columns and "longitude" in df.columns:
    lat_col, lon_col = "latitude", "longitude"
else:
    raise ValueError("Latitude/Longitude columns not found in the CSV.")

# 'class' usually already exists; if it comes as 'Class' in another dataset, standardize:
# df.rename(columns={"Class": "class"}, inplace=True)

spacex_df = df[["launch site", lat_col, lon_col, "class"]].copy()
spacex_df.rename(columns={lat_col: "Lat", lon_col: "Long", "launch site": "Launch Site"}, inplace=True)

spacex_df.head()

['Flight Number', 'Date', 'Time (UTC)', 'Booster Version', 'Launch Site', 'Payload', 'Payload Mass (kg)', 'Orbit', 'Customer', 'Landing Outcome', 'class', 'Lat', 'Long']


Launch Site        Lat       Long  class
0  CCAFS LC-40  28.562302 -80.577356      0
1  CCAFS LC-40  28.562302 -80.577356      0
2  CCAFS LC-40  28.562302 -80.577356      0
3  CCAFS LC-40  28.562302 -80.577356      0
4  CCAFS LC-40  28.562302 -80.577356      0

Now, you can take a look at what are the coordinates for each site.


In [66]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [68]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example,


In [71]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle.


Now, let's add a circle for each launch site in data frame `launch_sites`


*TODO:*  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [127]:
# Initial the map

# For each launch site, add a Circle object based on its coordinate (Lat, Long) values. In addition, add Launch site name as a popup label

# Initialize the map with NASA coordinates and initial zoom level 5
site_map = folium.Map(location=nasa_coordinate, zoom_start=5)

# Iterate over the 'launch_sites' DataFrame and add Circle and Marker for each point
for index, row in launch_sites_df.iterrows():
    coordinate = [row['Lat'], row['Long']]
    label = row['Launch Site']
    
    # Add circle
    folium.Circle(
        location=coordinate,
        radius=1000,
        color='#000000',
        fill=True
    ).add_child(folium.Popup(label)).add_to(site_map)
    
       # Add marker with styled HTML
    folium.map.Marker(
        location=coordinate,
        icon=folium.DivIcon(
            icon_size=(20,20),
            icon_anchor=(0,0),
            html=f'<div style="font-size: 12px; color:#d35400;"><b>{label}</b></div>',
        )
    ).add_to(site_map)

# Display the map
site_map

The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png">
</center>


### Now, you can explore the map by zoom-in/out the marked areas, and try to answer the following questions:

*   Are all launch sites in proximity to the Equator line?

No, not all launch sites are close to the Equator. Proximity to the Equator is advantageous for space launches due to the Earth’s rotational boost. However, based on the map:

	•	Launch sites in Florida (such as CCAFS and KSC) are relatively close to the Equator (around 28° N).
	•	Launch sites in California (like VAFB and SLC-4E) are located much farther north (above 34° N).

Therefore, while some launch sites are relatively near the Equator, not all of them are.

*   Are all launch sites in very close proximity to the coast?

  Yes, all launch sites are located very close to the coast. As shown on the map:
  
	•	CCAFS, KSC, and LC-39A are directly on the Atlantic coast.
	•	VAFB and SLC-4E are positioned along the Pacific coast.

Being near the coast allows rockets to launch over the ocean, minimizing risk to populated areas in case of a failure.

#### Also please try to explain your findings.


In [ ]:
# Task 2: Mark the success/failed launches for each site on the map


Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [77]:
spacex_df.tail(10)

Launch Site        Lat       Long  class
46    KSC LC-39A  28.573255 -80.646895      1
47    KSC LC-39A  28.573255 -80.646895      1
48    KSC LC-39A  28.573255 -80.646895      1
49  CCAFS SLC-40  28.563197 -80.576820      1
50  CCAFS SLC-40  28.563197 -80.576820      1
51  CCAFS SLC-40  28.563197 -80.576820      0
52  CCAFS SLC-40  28.563197 -80.576820      0
53  CCAFS SLC-40  28.563197 -80.576820      0
54  CCAFS SLC-40  28.563197 -80.576820      1
55  CCAFS SLC-40  28.563197 -80.576820      0

Next, let's create markers for all launch records.
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


In [79]:
marker_cluster = MarkerCluster()


*TODO:* Create a new column in `spacex_df` dataframe called `marker_color` to store the marker colors based on the `class` value


In [82]:

# Apply a function to check the value of `class` column

spacex_df['marker_color'] = spacex_df['class'].apply(lambda x: 'green' if x == 1 else 'red')

# If class=1, marker_color value will be green
# If class=0, marker_color value will be red

*TODO:* For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [85]:
# Add marker_cluster to current site_map
site_map.add_child(marker_cluster)

# for each row in spacex_df data frame
# create a Marker object with its coordinate
# and customize the Marker's icon property to indicate if this launch was successed or failed, 
# e.g., icon=folium.Icon(color='white', icon_color=row['marker_color']
# Add markers with custom color and popup
for index, record in spacex_df.iterrows():
    marker = folium.Marker(
        location=[record['Lat'], record['Long']],
        icon=folium.Icon(color='white', icon_color=record['marker_color']),
        popup=f"Launch Site: {record['Launch Site']}<br>Outcome: {'Success' if record['class'] == 1 else 'Failure'}"
    )
    marker_cluster.add_child(marker)


site_map

Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png">
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png">
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


In [ ]:
# TASK 3: Calculate the distances between a launch site to its proximities


Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [87]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


In [89]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

*TODO:* Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.


In [129]:
# find coordinate of the closet coastline
# e.g.,: Lat: 28.56367  Lon: -80.57163
# distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)

# --- Coastline point (use the corrected lat/lon you captured on the waterline) ---
# Example values you measured with MousePosition:
coastline_lat = 28.56847   # <-- put your final coastline latitude here
coastline_lon = -80.56902  # <-- put your final coastline longitude here

# Launch site coordinates (already defined earlier in the notebook)
# launch_site_lat = ...
# launch_site_lon = ...

# Recalculate great-circle distance to the coastline (in km)
distance_coastline = round(
    calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon), 2
)

# --- Coastline marker with distance label ---
marker_coast = folium.Marker(
    location=(coastline_lat, coastline_lon),
    icon=folium.DivIcon(
        html=f"""
        <div style="font-size: 12px; color: teal;">
            <b>{distance_coastline:.2f} km</b> to Coastline
        </div>
        """
    )
)
site_map.add_child(marker_coast)

# --- PolyLine from launch site to coastline (keep style minimal like the lab) ---
line_coast = folium.PolyLine(
    locations=[(launch_site_lat, launch_site_lon), (coastline_lat, coastline_lon)],
    color='teal',
    weight=2
)
site_map.add_child(line_coast)

# (Optional) tiny endpoints to highlight targets, matching your style
for (lat, lon, color) in [
    (coastline_lat, coastline_lon, 'teal'),
]:
    folium.CircleMarker((lat, lon), radius=5, color=color, fill=True, fill_opacity=0.9).add_to(site_map)

# Display map and numeric output (so the notebook shows the corrected value)
from IPython.display import display
display(site_map)
print({"Coastline (km)": distance_coastline})

{'Coastline (km)': 0.96}


In [131]:
# Create and add a folium.Marker on your selected closest coastline point on the map
# Display the distance between coastline point and launch site using the icon property 
# for example
# distance_marker = folium.Marker(
#    coordinate,
#    icon=DivIcon(
#        icon_size=(20,20),
#        icon_anchor=(0,0),
#        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance),
#        )
#    )

distance_marker = folium.Marker(
    location=[coastline_lat, coastline_lon],
    icon=folium.DivIcon(html=f'''
        <div style="font-size: 12px; color: red;">
            {round(distance_coastline, 2)} km
        </div>''')
)
site_map.add_child(distance_marker)

*TODO:* Draw a `PolyLine` between a launch site to the selected coastline point


In [133]:
# Create a `folium.PolyLine` object using the coastline coordinates and launch site coordinate
# lines=folium.PolyLine(locations=coordinates, weight=1)
coordinates = [[launch_site_lat, launch_site_lon], [coastline_lat, coastline_lon]]
lines = folium.PolyLine(locations=coordinates, weight=1)
site_map.add_child(lines)

Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png">
</center>


*TODO:* Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first


A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png">
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png">
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png">
</center>


In [ ]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site


In [137]:
# ============== NEAREST CITY ==============
city_lat  = 28.4012
city_lon  = -80.6041

distance_city = round(
    calculate_distance(launch_site_lat, launch_site_lon, city_lat, city_lon), 2
)

# Target marker with label "X.XX km to City"
marker_city = folium.Marker(
    location=(city_lat, city_lon),
    icon=folium.DivIcon(html=f"""
        <div style="font-size:12px; color: blue;">
            <b>{distance_city:.2f} km</b> to City
        </div>
    """)
)
site_map.add_child(marker_city)

# Line Site -> City
line_city = folium.PolyLine(
    locations=[(launch_site_lat, launch_site_lon), (city_lat, city_lon)],
    color='blue', weight=2
)
site_map.add_child(line_city)

# ============== NEAREST HIGHWAY ==============
highway_lat = 28.5623
highway_lon = -80.5706

distance_highway = round(
    calculate_distance(launch_site_lat, launch_site_lon, highway_lat, highway_lon), 2
)

marker_highway = folium.Marker(
    location=(highway_lat, highway_lon),
    icon=folium.DivIcon(html=f"""
        <div style="font-size:12px; color: green;">
            <b>{distance_highway:.2f} km</b> to Highway
        </div>
    """)
)
site_map.add_child(marker_highway)

line_highway = folium.PolyLine(
    locations=[(launch_site_lat, launch_site_lon), (highway_lat, highway_lon)],
    color='green', weight=2
)
site_map.add_child(line_highway)

# ============== NEAREST RAILWAY ==============
rail_lat = 28.57055
rail_lon = -80.58551

distance_rail = round(
    calculate_distance(launch_site_lat, launch_site_lon, rail_lat, rail_lon), 2
)

marker_rail = folium.Marker(
    location=(rail_lat, rail_lon),
    icon=folium.DivIcon(html=f"""
        <div style="font-size:12px; color: purple;">
            <b>{distance_rail:.2f} km</b> to Railway
        </div>
    """)
)
site_map.add_child(marker_rail)

line_rail = folium.PolyLine(
    locations=[(launch_site_lat, launch_site_lon), (rail_lat, rail_lon)],
    color='purple', weight=2
)
site_map.add_child(line_rail)

# ============== COASTLINE  ==============
# Use the values  captured at the WATERLINE (lat, lon):
coastline_lat = 28.56647   
coastline_lon = -80.56888  

distance_coast = round(
    calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon), 2
)

marker_coast = folium.Marker(
    location=(coastline_lat, coastline_lon),
    icon=folium.DivIcon(html=f"""
        <div style="font-size:12px; color: teal;">
            <b>{distance_coast:.2f} km</b> to Coastline
        </div>
    """)
)
site_map.add_child(marker_coast)

line_coast = folium.PolyLine(
    locations=[(launch_site_lat, launch_site_lon), (coastline_lat, coastline_lon)],
    color='teal', weight=2
)
site_map.add_child(line_coast)

# small points on the targets to highlight the endpoints:
for (lat, lon, color) in [
    (city_lat, city_lon, 'blue'),
    (highway_lat, highway_lon, 'green'),
    (rail_lat, rail_lon, 'purple'),
    (coastline_lat, coastline_lon, 'teal')
]:
    folium.CircleMarker((lat, lon), radius=5, color=color, fill=True, fill_opacity=0.9).add_to(site_map)

# Display the map and metrics in the output
from IPython.display import display
display(site_map)
print({
    "City (km)": distance_city,
    "Highway (km)": distance_highway,
    "Railway (km)": distance_rail,
    "Coastline (km)": distance_coast
})

{'City (km)': 18.22, 'Highway (km)': 0.62, 'Railway (km)': 1.18, 'Coastline (km)': 0.86}


After you plot distance lines to the proximities, you can answer the following questions easily:

*   Are launch sites in close proximity to railways?
*   Are launch sites in close proximity to highways?
*   Are launch sites in close proximity to coastline?
*   Do launch sites keep certain distance away from cities?

Also please try to explain your findings.

    •	Are launch sites in close proximity to railways?
   
Yes. The closest highway is only 0.62 km away, enabling efficient ground logistics and emergency access.

	•	Are launch sites in close proximity to highways?
    
Yes, the distance to the nearest highway is approximately 1.07 km, which shows that logistical access by road is efficient for vehicles and emergency services.

	•	Are launch sites in close proximity to coastline?
    
Yes. The coastline is about 0.96 km from the site. This aligns with best practices of launching near water to reduce overflight risk and enable offshore recovery.

    •	Do launch sites keep certain distance away from cities?
    
Yes. The nearest city is roughly 18.22 km away, confirming that launch pads are sited at a safe buffer from densely populated areas to minimize public risk.


# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


## Authors


[Pratiksha Verma](https://www.linkedin.com/in/pratiksha-verma-6487561b1/)


<!--## Change Log--!>


<!--| Date (YYYY-MM-DD) | Version | Changed By      | Change Description      |
| ----------------- | ------- | -------------   | ----------------------- |
| 2022-11-09        | 1.0     | Pratiksha Verma | Converted initial version to Jupyterlite|--!>


### <h3 align="center"> IBM Corporation 2022. All rights reserved. <h3/>
